### PR TITLE
Add CC Donation Modal to CC Global Components

### LINKED ISSUE #5: [Feature] Add CC Donation Modal to the CC Global Components
## Problem
We have previously used Google Tag Manager to define a donation appeal pop-up. Following from this (comment)[https://github.com/creativecommons/project_creativecommons.org/issues/186#issuecomment-984545850], 
@brylie suggested that:
> Alternatively, this might be a good addition to the cc-global-components collection, since it is just JavaScript, HTML, and CSS.

 - [x] Develop a component for the Donation Modal
 - [x] Set a cookie when the user closes the modal (by clicking close or outside of the modal)
 - [x] The cookie should expire after five days

## Description
Adding the Donation Modal to the CC Global Components is advantageous given that it will be distributed via CDN making it easier to deploy on any of the CC Web properties. 

## Implementation

- [x] I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by MuluhGodson**
## Fixes

Closes #5  by @MuluhGodson 

## Description
This PR adds the CC Global Donation modal as a Vue Component. 
- [x] Develop a component for the Donation Modal
- [x] Set a cookie when the user closes the modal (by clicking close or outside of the modal)
- [x] The cookie should expire after five days

## Technical details
This component sets a cookie and if the modal is close by the user, the modal doesn't display until the cookie expires.

## Tests
Pull the commit. 
Run `npm run serve` or serve the file `example.html`

## Screenshots
Donation modal with cookie expiry time set to 10 secsonds.
https://user-images.githubusercontent.com/40151808/144566703-aec0d8f5-6565-463f-8c3b-2330604557ac.mp4

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the *default* branch of the repository (`main` or `master`).
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

For the purposes of this DCO, "license" is equivalent to "license or public domain dedication," and "open source license" is equivalent to "open content license or public domain dedication."
<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: src/components/cc-donate.vue [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+      this.show = false;
+      this.transition = false;
+    }
+    console.log(userHasNotOptedOutFromDonationAppeal);
+  },
+  data() {
+    return {
+      show: true,
+      transition: true,
+      expiryPeriodInDays: 5,
```
Comment (MuluhGodson) [inline src/components/cc-donate.vue:80]: Here is where we enter the number of days the Cookie should be active for.

File: README.md [hunk 2]
```diff
@@ -16,6 +16,7 @@ The folder structure is as follows:
 - `dev/`: Use for serving the components for local testing during development.
 - `src`: This directory contains a sub-directory:
   - `components`: This is where our components code is found. For this project, we have three components:
+  - - `cc-donate.vue`: The component for the CC Donatiion modal.
```
Comment (brylie) [inline README.md]:
```suggestion
  - - `cc-donate.vue`: The component for the CC Donation modal.
```

File: example.html [hunk 3]
```diff
@@ -7,6 +7,11 @@
     <title>CC Global Components</title>
   </head>
   <body>
+    <!-- Div element use to mound the CC Donate modal-->
```
Comment (brylie) [inline example.html]:
```suggestion
    <!-- Show the CC Donate modal-->
```

File: src/components/cc-donate.vue [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+      </div>
+    </div>
+  </transition>
+</template>
+<script>
+import { defineComponent } from "vue";
+import Cookies from "js-cookie";
+
+export default defineComponent({
+  name: "cc-explore",
```
Comment (brylie) [inline src/components/cc-donate.vue]:
```suggestion
  name: "cc-donate",
```

File: src/components/cc-donate.vue [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+  mounted() {
+    var userHasOptedOutFromDonationAppeal = Cookies.get("optOutDonationAppeal");
+
+    if (userHasOptedOutFromDonationAppeal) {
+      this.show = false;
+      this.transition = false;
+    }
+    console.log("opted out?: ", userHasOptedOutFromDonationAppeal);
```
Comment (brylie) [inline src/components/cc-donate.vue]:
```suggestion
```

File: src/components/cc-donate.vue [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+  methods: {
+    closeModal() {
+      this.show = false;
+      this.optOutDonationAppeal();
+    },
+    optOutDonationAppeal() {
+      Cookies.set("optOutDonationAppeal", "true", {
+        expires: this.expiryPeriodInDays,
+      });
+      Cookies.set("_ga", undefined);
```
Comment (brylie) [inline src/components/cc-donate.vue]: I'm not sure we want to clear our Google Analytics cookie when users opt-out of the donation modal. GA seems like a separate concern.

File: src/components/cc-donate.vue [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+.modal-body {
+  padding: 2px 5px;
+  align-content: center;
+}
+
+.modal-footer {
+  padding: 2px 5px;
+  /*  background-color: #FEEDE9;*/
```
Comment (brylie) [inline src/components/cc-donate.vue]:
```suggestion
```

File: src/components/cc-donate.vue [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+  background: transparent;
+  z-index: 9999;
+  color: #fff;
+  transition: opacity 500ms;
+}
+.overlay-modal {
+  margin: 0px;
+  padding: 10px;
+  text-align: left;
+  /* height: 100px; */
```
Comment (brylie) [inline src/components/cc-donate.vue]:
```suggestion
```

File: src/components/cc-donate.vue [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+  background: #fff;
+  border-radius: 5px;
+  /*   width: auto; */
```
Comment (brylie) [inline src/components/cc-donate.vue]:
```suggestion
```

**Review by brylie — CHANGES_REQUESTED** [review]

**Review by brylie — COMMENTED** [review]
Overall, we should discuss with @annatuma whether we want to include the donation modal in our CC global widgets library. I.e., do we intend to solicit donations using this widget in the future, or is the Google Tag Manager approach sustainable enough? We may decide not to use a donation appeal overlay, but we need to discuss it with other stakeholders.

**Comment by MuluhGodson** [comment]
> Overall, we should discuss with @annatuma whether we want to include the donation modal in our CC global widgets library. I.e., do we intend to solicit donations using this widget in the future, or is the Google Tag Manager approach sustainable enough? We may decide not to use a donation appeal overlay, but we need to discuss it with other stakeholders.


That's a great idea. 
This PR is just a standby in case there is ever a need to include it into the CC Global components.

**Comment by brylie** [comment]
I converted the PR to draft to indicate it is still under consideration since we are unsure about using modal dialogues for future fundraising.

Otherwise, the code looks good.

**Comment by annatuma** [comment]
> Overall, we should discuss with @annatuma whether we want to include the donation modal in our CC global widgets library. I.e., do we intend to solicit donations using this widget in the future, or is the Google Tag Manager approach sustainable enough? We may decide not to use a donation appeal overlay, but we need to discuss it with other stakeholders.

Absolutely the right call Brylie - let's come back to this after discussing with stakeholders. Excellent to see the way we could implement this though is feasible.

**Pull request closed without merging**